How should semiologies be combined?

Here's some code that might clarify things. Note that I modified scores for semiologies **B** and **C** with respect to the previous table in this issue.

```python
In [1]: from mega_analysis.semiology import get_df_from_dicts, normalise_semiologies_df, combine_semiologies_df
   ...:
   ...: dicts = {
   ...:     'A': dict(X=0, Y=1, Z=1000),
   ...:     'B': dict(X=2, Y=1, Z=0),
   ...:     'C': dict(X=50, Y=50, Z=0),
   ...:     'D': dict(X=0, Y=12, Z=13),
   ...:     'E': dict(X=0, Y=0.1, Z=100),
   ...: }
   ...:
   ...: df = get_df_from_dicts(dicts)
   ...: df
Out[1]:
            X     Y     Z
Semiology
A           0   1.0  1000
B           2   1.0     0
C          50  50.0     0
D           0  12.0    13
E           0   0.1   100

In [2]: normalised_df = normalise_semiologies_df(df)
   ...: normalised_df
Out[2]:
               X           Y      Z
Semiology
A            0.0    0.100000  100.0
B          100.0   50.000000    0.0
C          100.0  100.000000    0.0
D            0.0   92.307692  100.0
E            0.0    0.100000  100.0

In [3]: combined_df = normalised_df.sum()
   ...: combined_df
Out[3]:
X    200.000000
Y    242.507692
Z    300.000000
dtype: float64

In [4]: combined_normalised_df = combined_df / combined_df.max() * 100
   ...: combined_normalised_df
Out[4]:
X     66.666667
Y     80.835897
Z    100.000000
dtype: float64
```

And with the first two semiologies only:

```python
In [1]: from mega_analysis.semiology import get_df_from_dicts, normalise_semiologies_df, combine_semiologies_df
   ...:
   ...: dicts = {
   ...:     'A': dict(X=0, Y=1, Z=1000),
   ...:     'B': dict(X=2, Y=1, Z=0),
   ...: }
   ...:
   ...: df = get_df_from_dicts(dicts)
   ...: df
Out[1]:
           X  Y     Z
Semiology
A          0  1  1000
B          2  1     0

In [2]: normalised_df = normalise_semiologies_df(df)
   ...: normalised_df
Out[2]:
               X     Y      Z
Semiology
A            0.0   0.1  100.0
B          100.0  50.0    0.0

In [3]: combined_df = normalised_df.sum()
   ...: combined_df
Out[3]:
X    100.0
Y     50.1
Z    100.0
dtype: float64

In [4]: combined_normalised_df = combined_df / combined_df.max() * 100
   ...: combined_normalised_df
Out[4]:
X    100.0
Y     50.1
Z    100.0
dtype: float64
```

--- a/mega_analysis/semiology.py
+++ b/mega_analysis/semiology.py
@@ -7,7 +7,6 @@ from typing import Dict, List, Optional
 import yaml
 import numpy as np
 import pandas as pd
-from sklearn.preprocessing import MinMaxScaler
 
 from .crosstab.file_paths import file_paths
 from .crosstab.hierarchy_class import Hierarchy
@@ -252,12 +251,13 @@ def get_possible_lateralities(term) -> List[Laterality]:
 
 def combine_semiologies(
         semiologies: List[Semiology],
-        normalise: bool = True,
+        normalise_method: Optional[str] = 'minmax',
+        normalise_zero_axis: bool = True,
         ) -> Dict[int, float]:
     df = get_df_from_semiologies(semiologies)
-    if normalise:
-        df = normalise_semiologies_df(df)
-    combined_df = combine_semiologies_df(df, normalise=normalise)
+    if normalise_method is not None:
+        df = normalise_semiologies_df(df, method=normalise_method)
+    combined_df = combine_semiologies_df(df, normalise=normalise_zero_axis)
     return combined_df
 
 
@@ -290,12 +290,20 @@ def get_df_from_dicts(
     return df
 
 
-def normalise_semiologies_df(semiologies_df: pd.DataFrame) -> pd.DataFrame:
+def normalise_semiologies_df(
+        semiologies_df: pd.DataFrame,
+        method='minmax',
+        ) -> pd.DataFrame:
     table = np.array(semiologies_df)
     data = table.T
-    scaler = MinMaxScaler((0, 100))
-    scaler.fit(data)
-    normalised = scaler.transform(data).T
+    if method == 'minmax':
+        from sklearn.preprocessing import MinMaxScaler
+        scaler = MinMaxScaler((0, 100))
+        scaler.fit(data)
+        normalised = scaler.transform(data).T
+    elif method == 'softmax':
+        from scipy.special import softmax
+        normalised = softmax(semiologies_df, axis=1)
     normalised_df = pd.DataFrame(
         normalised,
         columns=semiologies_df.columns,

--- a/mega_analysis/semiology.py
+++ b/mega_analysis/semiology.py
@@ -113,7 +113,7 @@ class Semiology:
             include_postictals: bool = False,
             possible_lateralities: Optional[List[Laterality]] = None,
             inverse_localising_values: bool = False,
-            ):
+    ):
         self.term = term
         self.symptoms_side = symptoms_side
         self.dominant_hemisphere = dominant_hemisphere
@@ -224,7 +224,7 @@ class Semiology:
                     )
         return all_combined_gifs
 
-    def get_num_datapoints_dict(self) -> Optional[dict]:
+    def get_num_datapoints_dict(self, method: str = 'proportions') -> Optional[dict]:
         query_lateralisation_result = self.query_lateralisation()
         if query_lateralisation_result is None:
             message = f'No results generated for semiology term "{self.term}"'
@@ -237,7 +237,15 @@ class Semiology:
             in zip(labels, patients)
             if num_datapoints > 0
         }
-        return num_datapoints_dict
+
+        if method != 'proportions':
+            return num_datapoints_dict
+
+        elif method == 'proportions':
+            total = sum(list(num_datapoints_dict.values()))
+            new_datatpoints = {
+                k: v*100/total for (k, v) in num_datapoints_dict.items()}
+            return new_datatpoints
 
 
 def get_possible_lateralities(term) -> List[Laterality]:
@@ -251,20 +259,21 @@ def get_possible_lateralities(term) -> List[Laterality]:
 
 def combine_semiologies(
         semiologies: List[Semiology],
-        normalise_method: Optional[str] = 'minmax',
+        normalise_method: Optional[str] = 'proportions',
         normalise_zero_axis: bool = True,
-        ) -> Dict[int, float]:
-    df = get_df_from_semiologies(semiologies)
+) -> Dict[int, float]:
+    df = get_df_from_semiologies(semiologies, normalise_method)
     if normalise_method is not None:
         df = normalise_semiologies_df(df, method=normalise_method)
-    combined_df = combine_semiologies_df(df, normalise=normalise_zero_axis)
+    combined_df = combine_semiologies_df(
+        df, method=normalise_method, normalise=normalise_zero_axis)
     return combined_df
 
 
-def get_df_from_semiologies(semiologies: List[Semiology]) -> pd.DataFrame:
+def get_df_from_semiologies(semiologies: List[Semiology], method: str = 'proportions') -> pd.DataFrame:
     num_datapoints_dicts = {}
     for semiology in semiologies:
-        num_datapoints_dict = semiology.get_num_datapoints_dict()
+        num_datapoints_dict = semiology.get_num_datapoints_dict(method=method)
         if num_datapoints_dict is None:
             message = (
                 f'Information for semiology term "{semiology.term}"'
@@ -280,7 +289,7 @@ def get_df_from_semiologies(semiologies: List[Semiology]) -> pd.DataFrame:
 
 def get_df_from_dicts(
         semiologies_dicts: Dict[str, Dict[int, float]],
-        ) -> pd.DataFrame:
+) -> pd.DataFrame:
     records = []
     semiologies_dicts = copy.deepcopy(semiologies_dicts)
     for term, num_datapoints_dict in semiologies_dicts.items():
@@ -292,8 +301,11 @@ def get_df_from_dicts(
 
 def normalise_semiologies_df(
         semiologies_df: pd.DataFrame,
-        method='minmax',
-        ) -> pd.DataFrame:
+        method='proportions',
+) -> pd.DataFrame:
+    if method == 'proportions':
+        return semiologies_df
+
     table = np.array(semiologies_df)
     data = table.T
     if method == 'minmax':
@@ -314,11 +326,15 @@ def normalise_semiologies_df(
 
 def combine_semiologies_df(
         df: pd.DataFrame,
+        method: str = 'proportions',
         normalise: bool = True,
-        ) -> Dict[int, float]:
-    combined_df = df.sum()
-    if normalise:
-        combined_df = combined_df / combined_df.max() * 100
+) -> Dict[int, float]:
+    if method == 'proportions':
+        combined_df = df.mean(axis=0)
+    else:
+        combined_df = df.sum()
+        if normalise:
+            combined_df = combined_df / combined_df.max() * 100
     combined_df = pd.DataFrame(combined_df).T
     combined_df.index = ['Score']
     return combined_df

--- a/slicer/SemiologyVisualisation.py
+++ b/slicer/SemiologyVisualisation.py
@@ -387,11 +387,14 @@ class SemiologyVisualisationWidget(ScriptedLoadableModuleWidget):
         normalisationLayout = qt.QHBoxLayout(normalisationGroupBox)
 
         self.minmaxRadioButton = qt.QRadioButton('Rescaling')
-        self.minmaxRadioButton.setChecked(True)
         normalisationLayout.addWidget(self.minmaxRadioButton)
 
-        self.softmaxRadioButton = qt.QRadioButton('Soft argmax')
+        self.softmaxRadioButton = qt.QRadioButton('Softmax')
         normalisationLayout.addWidget(self.softmaxRadioButton)
+
+        self.proportionsRadioButton = qt.QRadioButton('Proportions')
+        self.proportionsRadioButton.setChecked(True)
+        normalisationLayout.addWidget(self.proportionsRadioButton)
 
         return advancedTabWidget
 
@@ -589,7 +592,13 @@ class SemiologyVisualisationWidget(ScriptedLoadableModuleWidget):
         else:
             with messageContextManager('Querying mega_analysis module...'):
                 try:
-                    dataFrame = get_df_from_semiologies(semiologies)
+                    if self.minmaxRadioButton.isChecked():
+                        method = 'minmax'
+                    elif self.softmaxRadioButton.isChecked():
+                        method = 'softmax'
+                    elif self.proportionsRadioButton.isChecked():
+                        method = 'proportions'
+                    dataFrame = get_df_from_semiologies(semiologies, method=method)
                     scores = Scores(dataFrame)
                     cache[hashedQuery] = scores.toDict()
                     self.logic.writeCache(cache)
@@ -711,15 +720,17 @@ class SemiologyVisualisationWidget(ScriptedLoadableModuleWidget):
                 method = 'minmax'
             elif self.softmaxRadioButton.isChecked():
                 method = 'softmax'
+            elif self.proportionsRadioButton.isChecked():
+                method = 'proportions'
             normalisedDataFrame = normalise_semiologies_df(
                 semiologiesDataFrame,
                 method=method,
             )
             combinedDataFrame = combine_semiologies_df(
-                normalisedDataFrame, normalise=True)
+                normalisedDataFrame, method=method, normalise=True)
         else:
             combinedDataFrame = combine_semiologies_df(
-                semiologiesDataFrame, normalise=False)
+                semiologiesDataFrame, method=method, normalise=False)
 
         if self.logic.dataFrameIsEmpty(combinedDataFrame):
             slicer.util.errorDisplay('The combined results are empty')
@@ -825,6 +836,7 @@ class SemiologyVisualisationWidget(ScriptedLoadableModuleWidget):
         else:
             self.inverseLocalisingCheckBox.setChecked(False)
             self.inverseLocalisingCheckBox.setEnabled(False)
+
 
 class SemiologyVisualisationLogic(ScriptedLoadableModuleLogic):
 

--- a/slicer/SemiologyVisualisation.py
+++ b/slicer/SemiologyVisualisation.py
@@ -368,13 +368,30 @@ class SemiologyVisualisationWidget(ScriptedLoadableModuleWidget):
         advancedTabWidget = qt.QWidget()
         advancedTabLayout = qt.QVBoxLayout(advancedTabWidget)
 
+        # Cache
+        cacheGroupBox = qt.QGroupBox('Cache')
+        advancedTabLayout.addWidget(cacheGroupBox)
+        cacheLayout = qt.QVBoxLayout(cacheGroupBox)
+
         self.useCacheCheckBox = qt.QCheckBox('Use cached queries if available')
         self.useCacheCheckBox.setChecked(False)
-        advancedTabLayout.addWidget(self.useCacheCheckBox)
+        cacheLayout.addWidget(self.useCacheCheckBox)
 
         self.clearCacheButton = qt.QPushButton('Clear cache')
         self.clearCacheButton.clicked.connect(self.logic.clearCache)
-        advancedTabLayout.addWidget(self.clearCacheButton)
+        cacheLayout.addWidget(self.clearCacheButton)
+
+        # Normalisation
+        normalisationGroupBox = qt.QGroupBox('Normalisation function')
+        advancedTabLayout.addWidget(normalisationGroupBox)
+        normalisationLayout = qt.QHBoxLayout(normalisationGroupBox)
+
+        self.minmaxRadioButton = qt.QRadioButton('Rescaling')
+        self.minmaxRadioButton.setChecked(True)
+        normalisationLayout.addWidget(self.minmaxRadioButton)
+
+        self.softmaxRadioButton = qt.QRadioButton('Soft argmax')
+        normalisationLayout.addWidget(self.softmaxRadioButton)
 
         return advancedTabWidget
 
@@ -690,8 +707,14 @@ class SemiologyVisualisationWidget(ScriptedLoadableModuleWidget):
         semiologiesDataFrame = self.getSemiologiesDataFrameFromGUI()
         normalise = len(semiologiesDataFrame) > 1
         if normalise:
+            if self.minmaxRadioButton.isChecked():
+                method = 'minmax'
+            elif self.softmaxRadioButton.isChecked():
+                method = 'softmax'
             normalisedDataFrame = normalise_semiologies_df(
-                semiologiesDataFrame)
+                semiologiesDataFrame,
+                method=method,
+            )
             combinedDataFrame = combine_semiologies_df(
                 normalisedDataFrame, normalise=True)
         else:

--- a/slicer/SemiologyVisualisation.py
+++ b/slicer/SemiologyVisualisation.py
@@ -714,7 +714,7 @@ class SemiologyVisualisationWidget(ScriptedLoadableModuleWidget):
             return
 
         semiologiesDataFrame = self.getSemiologiesDataFrameFromGUI()
-        normalise = len(semiologiesDataFrame) > 1
+        normalise = True  # len(semiologiesDataFrame) > 1
         if normalise:
             if self.minmaxRadioButton.isChecked():
                 method = 'minmax'

--- a/slicer/SemiologyVisualisation.py
+++ b/slicer/SemiologyVisualisation.py
@@ -972,6 +972,9 @@ class SemiologyVisualisationLogic(ScriptedLoadableModuleLogic):
             if colorNode.GetName() not in COLORMAPS:
                 slicer.mrmlScene.RemoveNode(colorNode)
 
+    def getPythonConsoleWidget(self):
+        return slicer.util.mainWindow().pythonConsole().parent()
+
     def installRepository(self):
         repoDir = Path(__file__).parent.parent
         import sys
@@ -983,9 +986,13 @@ class SemiologyVisualisationLogic(ScriptedLoadableModuleLogic):
                     import mega_analysis
             except ImportError:
                 requirementsPath = repoDir / 'requirements.txt'
+                console = self.getPythonConsoleWidget()
+                pythonVisible = console.visible
+                console.setVisible(True)
                 slicer.util.pip_install(
                     f'-r {requirementsPath}'
                 )
+                console.setVisible(pythonVisible)
             import matplotlib
             matplotlib.use('agg')
 

--- a/tests/test_dummy_data.py
+++ b/tests/test_dummy_data.py
@@ -303,7 +303,7 @@ class TestDummyDataDummyDictionary(unittest.TestCase):
                 == lat_not_loc_gifsclean.shape[0])
 
         # test MTG on right 155 gif # gives 1:
-        heatmap = patient.get_num_datapoints_dict()
+        heatmap = patient.get_num_datapoints_dict(method='minmax')
         assert 156 not in heatmap  # left
         assert heatmap[155] == 1  # right
 
@@ -354,7 +354,7 @@ class TestDummyDataDummyDictionary(unittest.TestCase):
                 == lat_not_loc_gifsclean.shape[0])
 
         # check that latnotloc gives 1 and latandloc adds zero to right MTG GIF #155
-        heatmap = patient.get_num_datapoints_dict()
+        heatmap = patient.get_num_datapoints_dict(method='minmax')
         assert heatmap[155] == 1  # right
 
     def test_latexceedsloc_3(self):
@@ -371,7 +371,7 @@ class TestDummyDataDummyDictionary(unittest.TestCase):
         patient.data_frame = self.df
 
         # (part 1) test latexceedsloc alone using norm_ratio/oddsratio method of Q_L:
-        heatmap = patient.get_num_datapoints_dict()
+        heatmap = patient.get_num_datapoints_dict(method='minmax')
         assert heatmap[156] == 200.0
         assert heatmap[155] == 300.0
 
@@ -443,7 +443,7 @@ class TestDummyDataDummyDictionary(unittest.TestCase):
         assert (lat_allgifs.loc[165, 'pt #s'] == 0)
 
         lat_allgifs = lat_allgifs.loc[lat_allgifs['pt #s'] != 0, :]
-        SVT_output = patient.get_num_datapoints_dict()
+        SVT_output = patient.get_num_datapoints_dict(method='minmax')
         assert SVT_output == dict((lat_allgifs)['pt #s'])
 
 

